File: src/schemas/portfolio_instruction.py
"""User-facing projection for internal portfolio action axes."""

from __future__ import annotations

from typing import Literal


HoldingInstruction = Literal["add", "hold", "reduce", "exit", "insufficient"]

_POSITION_ACTIONS = frozenset({"hold", "reduce", "exit"})
_INCREMENTAL_ACTIONS = frozenset({"add_in_batches", "wait", "no_add"})


def project_holding_instruction(
    *,
    position_action: str,
    incremental_action: str,
    blocked: bool,
) -> HoldingInstruction:
    """Return one plain instruction without changing either stored action axis."""

    if (
        position_action not in _POSITION_ACTIONS
        or incremental_action not in _INCREMENTAL_ACTIONS
    ):
        raise ValueError("invalid_portfolio_action")
    if position_action in {"reduce", "exit"} and incremental_action == "add_in_batches":
        raise ValueError("contradictory_portfolio_actions")
    if blocked:
        return "insufficient"
    if position_action == "reduce":
        return "reduce"
    if position_action == "exit":
        return "exit"
    if incremental_action == "add_in_batches":
        return "add"
    return "hold"

File: apps/dsa-web/src/utils/portfolioInstruction.ts
export type HoldingInstruction = 'add' | 'hold' | 'reduce' | 'exit' | 'insufficient';
export type SelectableHoldingInstruction = Exclude<HoldingInstruction, 'insufficient'>;

export type PositionAction = 'hold' | 'reduce' | 'exit';
export type IncrementalAction = 'add_in_batches' | 'wait' | 'no_add';

type InternalAxes = {
  positionAction: PositionAction;
  incrementalAction: IncrementalAction;
};

export function selectableInstructionFromAxes(
  positionAction: string,
  incrementalAction: string,
): SelectableHoldingInstruction {
  if (positionAction === 'reduce') return 'reduce';
  if (positionAction === 'exit') return 'exit';
  if (incrementalAction === 'add_in_batches') return 'add';
  return 'hold';
}

export function instructionToAxes(
  instruction: HoldingInstruction,
  currentPositionAction: string,
  currentIncrementalAction: string,
): InternalAxes {
  if (instruction === 'insufficient') {
    throw new Error('insufficient_instruction_not_selectable');
  }
  if (instruction === 'add') {
    return { positionAction: 'hold', incrementalAction: 'add_in_batches' };
  }
  if (instruction === 'reduce') {
    return { positionAction: 'reduce', incrementalAction: 'no_add' };
  }
  if (instruction === 'exit') {
    return { positionAction: 'exit', incrementalAction: 'no_add' };
  }
  if (
    currentPositionAction === 'hold'
    && (currentIncrementalAction === 'wait' || currentIncrementalAction === 'no_add')
  ) {
    return {
      positionAction: 'hold',
      incrementalAction: currentIncrementalAction,
    };
  }
  return { positionAction: 'hold', incrementalAction: 'no_add' };
}
